src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -103,6 +108,22 @@ public static boolean isStringSafe(@CheckForNull String input, @CheckForNull Env
                 : true;
     }
 
+    /** Restrict stored arguments to a reasonable subset of types so we don't retain totally arbitrary objects
+     *  in memory. Generally we aim to allow storing anything that maps correctly to an {@link UninstantiatedDescribable}
+     *  But we may allow a few extras doesn't address, if they're safe to store.
+     */
+    boolean isStoreableType(Object ob) {
+        if (ob == null) {
+            return true;
+        } else if (ob instanceof CharSequence || ob instanceof Number || ob instanceof Boolean
+                || ob instanceof Map || ob instanceof List || ob instanceof UninstantiatedDescribable
```

Comment:
Huh?  Map is used to represent an exploded class.  Anyway, we would need to pass that through anyway as legal since we're using map to hold the exploded arguments (or using `#toMap` on `UninstantiatedDescribable`s).

Reply:
I should have been clearer. `Map` is used to _represent_ `UninstantiatedDescribable` but it is not honored as a supported data type for input per se. For example, you cannot declare a step (or any `Describable`) parameter to be of type `Map<String, Integer>`.

Reply:
IOW you _can_ decide to record such anomalous arguments, provided it seems safe to do so, but you _need not_ record anything outside of the officially supported data types.

`ChoiceParameterDefinition` and its usage of `List<String>` is a special mess. See https://github.com/jenkinsci/jenkins/pull/3014. The proper way to solve this would have been to declare `choices` to be of type `List<Choice>` where

```java
public static class Choice extends AbstractDescribableImpl<Choice> {
    public final String value;
    @DataBoundConstructor public Choice(String value) {this.value = value;}
    // optionally:
    @DataBoundSetter public String displayName;
    @Symbol("choice")
    @Extension public static class DescriptorImpl extends Descriptor<Choice> {}
}
```

which would allow use of normal `config.xml` idioms and also allow Pipeline usage like

```groovy
properties([parameters([choice(name: 'mode', choices: [choice('full'), choice('partial')])])])
```

without any special casing except a `newInstance` override to handle compatibility with scripts passing the old syntax

```groovy
choices: 'full\npartial'
```